### DISCUSSION TITLE
Julia Interface for User-Defined Functions

### DISCUSSION THREAD

**Discussion by tqml** [duckdb/duckdb#13176] [General]
Hey!

I think this feature is pretty cool (as DuckDB in general) and since I have some extra time in the next two months I would like to help out with the Julia implementation of this feature.

I envision the Julia API to look like the following:

```julia
using DuckDB
import DuckDB: execute, @create_scalar_function, register_scalar_function
db = DuckDB.DB(":memory:")
my_sum = (a,b) -> a+b

# use of a macro to generate specialized wrapper code
fun = @create_scalar_function "my_sum" [Int64, Int64] Int64 my_sum
register_scalar_function(db, fun)

execute(db, "SELECT my_sum(1,2)")

# way to remove this as well
unregister_scalar_function(db, "my_sum")
```

I created a first example here [scalar_function.jl](https://github.com/tqml/duckdb/blob/julia/scalar-function/tools/juliapkg/src/scalar_function.jl). It is based on the existing implementation of table functions with the addition to have a macro that can implement a specialized wrapper function the correct data types. The implementation requires the most recent build of duckdb (https://github.com/duckdb/duckdb/discussions/4874#discussioncomment-10101689) but seems to work so far. 

Let me know what you think!

**Comment by dwinkler1** [comment]
I can't open the link but the idea sounds great! DuckDB and Julia are a great combo for DS.

**Reply by tqml** [reply]
The DuckDB and Julia Combo is amazing!

Not sure about the links, they seem to work for me 🤔

I want to polish it a bit and create a draft PR

**Reply by dwinkler1** [reply]
Thank you for your work! Links are working now, possibly my connection earlier.

**Reply by drizk1** [reply]
Also can't see the work

**Comment by drizk1** [comment]
@tqml Sorry I'm just seeing this. 

I just put in the [following pr](https://github.com/duckdb/duckdb/pull/14024) and it was merged. 

That being said your interface is much smoother .. I way prefer it. Is it possible for you to still add it ?

We could possible figure out how to do [aggregate](https://github.com/duckdb/duckdb/blob/main/test/api/capi/capi_aggregate_functions.cpp) functions too

**Reply by tqml** [reply]
Hey @drizk1,

no worries I was quite busy last month and couldnt really continue working on it.

Yeah, I definitely would like to add is as well :) 
My idea was to create a macro as convenience wrapper around the raw API so that all the data type conversions, registrations etc. is taken care off. Also in the Julia spirit, i was hoping a macro could create faster static code and avoids unnecessary dynamic de-referencing.

My issue when I wrote it is, was that I couldn't get it to work with Strings (or structs), do you have any pointers there?

I will polish up the code and create a draft/wip PR

**Reply by drizk1** [reply]
The macro wouldn't work for strings or structs ? Or the scalar function? 

It's tough to say without being able to see the code. Would you be able to share it somehow? And I can try it out

**Reply by tqml** [reply]
The implementation currently produces a segmentation fault when one of the arguments has a data type other than Int/Float, which is most likely due to missing string casting.

In my understanding, the vector in the data chunk is a vector of (smart) string pointers that would need conversion to the Julia native String type to be used in Julia (and of course also the other way round if the result should be of type String).
With other data types I think it should be similar but haven't looked at it yet.

**Reply by drizk1** [reply]
Oh I see for the one that was recently merged?

**Comment by tecosaur** [comment]
A quick syntax comment, it seems to me that

```julia
fun = @create_scalar_function "my_sum" [Int64, Int64] Int64 my_sum
```

Could be alternatively expressed as

```julia
fun = @create_scalar_function mysum(::Int64, ::Int64)::Int64 as "my_sum"
```

which reads a bit better to me.

**Reply by Tishj** [reply]
Agreed, that reads a bit more intuitively 👍

**Reply by tecosaur** [reply]
You could probably even drop the quotes in `as "my_sum"` and also make it default to same function name. I.e. allow you to omit `as my_sum` when you don't want to change the function name:

```julia
fun = @create_scalar_function my_sum(::Int64, ::Int64)::Int64
```

**Comment by tqml** [comment]
This feature with the proposed interface was merged in PR #15430  :)
